### ISSUE TITLE
I want to implement upload using graphql + next.js. Can I get help?

### ISSUE THREAD

**Issue by namiomiru** [jaydenseric/graphql-upload#216]
I want to implement upload using graphql + next.js. Can I get help?

https://github.com/prisma/prisma-examples/tree/master/typescript/graphql-nextjs

**Comment by jaydenseric** [comment]
GitHub issues is not the best place to get general advice or help with a project, but here is a tip: Keep your GraphQL API and whatever apps use it as separate projects with seperate servers that deploy separately. They can be in a monorepo, or separate repos, but make sure they have separate `package.json` files / dependencies. Ideally your API lasts a long time; with a separation you can make multiple apps using the same API and throw away and replace old apps without getting tangled up in the API.

This also reduces the chance that a compromised dependency used in your app will mess with your backend and user data.

If you use Apollo, there is an example of file uploads with a separate API and Next.js app:

https://github.com/jaydenseric/apollo-upload-examples

Personally, I use [`graphql-api-koa`](https://github.com/jaydenseric/graphql-api-koa) which is easy to use with `graphql-upload` if you follow the readme instructions, and for Next.js apps I use [`graphql-react`](https://github.com/jaydenseric/graphql-react)/[`next-graphql-react`](https://github.com/jaydenseric/next-graphql-react), which supports file uploads out of the box, no special setup needed (unlike Apollo Client).

**Comment by stomvi** [comment]
TL;DR, if you are using `graphql-uplaod` + `apollo-server-micro` + next.js, you might not pass the `GraphQLUpload.js` check and get an `Upload value invalid` error.

Just and an custom alias config into the `next.config.js`
[https://nextjs.org/docs/api-reference/next.config.js/introduction](https://nextjs.org/docs/api-reference/next.config.js/introduction)

```javascript
config.resolve.alias['graphql-upload'] =
      'the/path/to/your/project/node_modules/graphql-upload/public/index.js';
```

I guess the error is caused by the module splitting of `next.js`.  `Upload` class was imported multiple times, even the `resolutions` solution from (#185)[https://github.com/jaydenseric/graphql-upload/issues/185#issuecomment-758822484] doesn't work for me.

**Comment by vincenujib** [comment]
hi @stomvi, i am having that error `Upload value invalid` but the new version of graphql-upload 15.0 doesn't have the node_modules/graphql-upload/public/index.js anymore. What do you think is the solution?

**Comment by jaydenseric** [comment]
@vincenujib see https://github.com/jaydenseric/graphql-upload/issues/305, and https://github.com/jaydenseric/graphql-upload/issues/305#issuecomment-1134736158 for an explanation and https://github.com/jaydenseric/graphql-upload/issues/305#issuecomment-1135285811 for the new way to write the import.

**Comment by vincenujib** [comment]
> @vincenujib see #305, and [#305 (comment)](https://github.com/jaydenseric/graphql-upload/issues/305#issuecomment-1134736158) for an explanation and [#305 (comment)](https://github.com/jaydenseric/graphql-upload/issues/305#issuecomment-1135285811) for the new way to write the import.

Hi @jaydenseric , i checked that and got this error. `Variable \"$file\" got invalid value { path: \"279453655_2593028377495378_4749023162893893740_n.jpg\" }; Upload value invalid."`

**Comment by jaydenseric** [comment]
@vincenujib that problem is not related to this specific issue. There are past issues (either here, or in https://github.com/jaydenseric/apollo-upload-client) about `react-dropzone` you can look at.

**Comment by vincenujib** [comment]
All good now @jaydenseric.

Solve this guys by adding `createUploadLink`.

`import { createUploadLink } from 'apollo-upload-client';

const client = new ApolloClient({
  uri: '/api/graphql',
  cache: new InMemoryCache(),
  link: createUploadLink({ uri: '/api/graphql' }),
});`

Then add `processRequest` to your `ApolloServer`

`import processRequest from 'graphql-upload/processRequest.js';

const contentType = req.headers['content-type'];
if (contentType && contentType.startsWith('multipart/form-data')) {
  req.filePayload = await processRequest(req, res);
}`

**Comment by hafaiedhmehdy** [comment]
> All good now @jaydenseric.
> 
> Solve this guys by adding `createUploadLink`.
> 
> ```
> import { createUploadLink } from 'apollo-upload-client';
> 
> const client = new ApolloClient({
>   uri: '/api/graphql',
>   cache: new InMemoryCache(),
>   link: createUploadLink({ uri: '/api/graphql' }),
> });
> ```
> 
> Then add `processRequest` to your `ApolloServer`
> 
> ```
> import processRequest from 'graphql-upload/processRequest.js';
> 
> const contentType = req.headers['content-type'];
> if (contentType && contentType.startsWith('multipart/form-data')) {
>   req.filePayload = await processRequest(req, res);
> }
> ```

How's this possible with Next.js App Router? It's definitely not the way, the underlying of processRequest, busboy doesn't support Next.js new API at all. See here: https://github.com/meabed/graphql-upload-ts/issues/309#issue-2323584096

If anyone has a suggestion, don't hesitate!

**Comment by rohitmangoit** [comment]
I'm getting this error 
Payload : 

operations: {"operationName":"organizationLogo","variables":{"organizationId":"4c5a1413-9fcc-4d27-97d5-2d0a8f4b02a7","file":null},"query":"mutation organizationLogo($organizationId: String!, $file: Upload!) {\n  organizationLogo(organizationId: $organizationId, file: $file) {\n    organizationId\n    file\n    __typename\n  }\n}"}
map: {"1":["variables.file"]}
1: (binary)

[
    {
        "message": "Context creation failed: Unexpected end of form",
        "extensions": {
            "code": "INTERNAL_SERVER_ERROR",
            "stacktrace": [
                "Error: Unexpected end of form",
                "    at Multipart._final (/home/rohit/Desktop/exch-api/exchange-next-web/node_modules/busboy/lib/types/multipart.js:588:17)",
                "    at callFinal (node:internal/streams/writable:698:12)",
                "    at prefinish (node:internal/streams/writable:710:7)",
                "    at finishMaybe (node:internal/streams/writable:720:5)",
                "    at Writable.end (node:internal/streams/writable:634:5)",
                "    at onend (node:internal/streams/readable:716:10)",
                "    at process.processTicksAndRejections (node:internal/process/task_queues:77:11)"
            ]
        }
    }
]

Anyone please help.

**Comment by jaydenseric** [comment]
@rohitmangoit this issue is not the right place to ask for help, but check out the “Tips” section in the readme:

https://github.com/jaydenseric/graphql-upload/tree/v16.0.2?tab=readme-ov-file#tips

Particularly this tip:

> Promisify and await file upload streams in resolvers or the server will send a response to the client before uploads are complete, causing a disconnect.

If it's not that, investigate if something is causing the client to abort the request before it's finished.

**Comment by hafaiedhmehdy** [comment]
> I'm getting this error Payload :
> 
> operations: {"operationName":"organizationLogo","variables":{"organizationId":"4c5a1413-9fcc-4d27-97d5-2d0a8f4b02a7","file":null},"query":"mutation organizationLogo($organizationId: String!, $file: Upload!) {\n organizationLogo(organizationId: $organizationId, file: $file) {\n organizationId\n file\n __typename\n }\n}"} map: {"1":["variables.file"]} 1: (binary)
> 
> [ { "message": "Context creation failed: Unexpected end of form", "extensions": { "code": "INTERNAL_SERVER_ERROR", "stacktrace": [ "Error: Unexpected end of form", " at Multipart._final (/home/rohit/Desktop/exch-api/exchange-next-web/node_modules/busboy/lib/types/multipart.js:588:17)", " at callFinal (node:internal/streams/writable:698:12)", " at prefinish (node:internal/streams/writable:710:7)", " at finishMaybe (node:internal/streams/writable:720:5)", " at Writable.end (node:internal/streams/writable:634:5)", " at onend (node:internal/streams/readable:716:10)", " at process.processTicksAndRejections (node:internal/process/task_queues:77:11)" ] } } ]
> 
> Anyone please help.

Hi @rohitmangoit, I'm working on this, but it's a bit complicated. The main issue is the one-time read limitation of the request object in Next.js +13, and Busboy isn't currently compatible with this new version either.

**Comment by hafaiedhmehdy** [comment]
Well, my solution for handling file uploads in Next.js using Apollo Server is now available! It's an alpha version, and I'm welcoming any feedback or criticism to improve it further. It's missing acknowledgements, so I welcome people to add themselves before I do, as I'm really busy. You can check it out here: [graphql-upload-nextjs](https://github.com/hafaiedhmehdy/graphql-upload-nextjs).
